synapse/handlers/send_email.py

```diff
@@ -109,23 +115,13 @@ def build_sender_factory(**kwargs: Any) -> ESMTPSenderFactory:
         # set to enable TLS.
         factory = build_sender_factory(hostname=smtphost if enable_tls else None)
 
+    endpoint = HostnameEndpoint(
+        reactor, smtphost, smtpport, timeout=30, bindAddress=None
+    )
     if force_tls:
-        reactor.connectSSL(
-            smtphost,
-            smtpport,
-            factory,
-            optionsForClientTLS(smtphost),
-            timeout=30,
-            bindAddress=None,
-        )
-    else:
-        reactor.connectTCP(
-            smtphost,
-            smtpport,
-            factory,
-            timeout=30,
-            bindAddress=None,
-        )
+        factory = ESMTPTLSClientFactory(optionsForClientTLS(smtphost), True, factory)
```

Comment:
I'm failing to find examples of using `TLSMemoryBIOFactory` with `HostnameEndpoint` this way; it seems plausible, but I'm not confident in it.

Reply:
First of all: thanks a lot for the thorough review! Lovely to have issues caught before they land :-).

So, regarding this: you are right that this is something that we see seldom explicitly (we know Twisted's docs can use some help in improving them), but we also use it a lot actually:

`PosixreactorBase.connectSSL` contains [this line](https://github.com/twisted/twisted/blob/e2e113c04ad1ff991f737a7d81a3139512ddc2ab/src/twisted/internet/posixbase.py#L374):

```
    def connectSSL(
        self, host, port, factory, contextFactory, timeout=30, bindAddress=None
    ):
        if tls is not None:
            tlsFactory = tls.TLSMemoryBIOFactory(contextFactory, True, factory)
            return self.connectTCP(host, port, tlsFactory, timeout, bindAddress)
```

If we think about it, that's precisely what we are doing by passing `TLSMemoryBioFactory` this way, given that:

`HostnameEndpoint.connect` contains [this line](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/endpoints.py#L1050):

```
    def connect(self, protocolFactory):
      [...]
                eachAttempt = endpoint.connect(protocolFactory)
```

Where `endpoint` has already been resolved to one of the candidate hosts (directly as IPv{4,6} address).

So, all in all, what we are doing is: instead of using `reactor.connectSSL`, which adds the `TLSMemoryBIOFactory` before actually calling `connectTCP` (which has the host-resolving issues that cause this class of bugs), we use `HostnameEndpoint` and directly use `TLSMemoryBIOFactory` as the wrapper protocol factory.

Hope that makes sense code and theory-wise.

When it comes to actually testing it: it is working :-). I did validate the possible paths with a couple homeservers I control.